dev_tools/oneapi/convert_source.sh

```diff
@@ -0,0 +1,429 @@
+#   GTEST_HEADER_DIR: the gtest header folder. The default is "${ROOT_BUILD_DIR}/third_party_gtest/src/googletest/include"
+#   CLANG_FORMAT: the clang-format exec. The default is "clang-format"
+CURRENT_DIR="$( pwd )"
+cd "$( dirname "${BASH_SOURCE[0]}" )"
+SCRIPT_DIR="$( pwd )"
+ROOT_DIR="${ROOT_DIR:="${CURRENT_DIR}"}"
+BUILD_DIR="${BUILD_DIR:="build"}"
+ROOT_BUILD_DIR="${ROOT_BUILD_DIR:="${ROOT_DIR}/${BUILD_DIR}"}"
+CUDA_HEADER_DIR="${CUDA_HEADER_DIR}"
+GTEST_HEADER_DIR="${GTEST_HEADER_DIR:="${ROOT_BUILD_DIR}/third_party/gtest/src/googletest/include"}"
+CLANG_FORMAT=${CLANG_FORMAT:="clang-format"}
+if [[ "${VERBOSE}" == 1 ]]; then
+    echo "the current setting is "
+    echo "CURRENT_DIR ${CURRENT_DIR}"
+    echo "SCRIPT_DIR ${SCRIPT_DIR}"
+    echo "ROOT_DIR ${ROOT_DIR}"
+    echo "ROOT_BUILD_DIR ${ROOT_BUILD_DIR}"
+    echo "GTEST_HEADER_DIR ${GTEST_HEADER_DIR}"
+    echo "CUDA_HEADER_DIR ${CUDA_HEADER_DIR}"
+    echo "ROOT_BUILD_DIR ${ROOT_BUILD_DIR}"
+    echo "CLANG_FORMAT ${CLANG_FORMAT}"
+fi
+if [[ "${CUDA_HEADER_DIR}" == "" ]]; then
+    echo "Please set the CUDA_HEADER_DIR in environment"
+    exit 1
+fi
+# move to working_directory
+cd working_directory
+
+
+KERNEL_SYNTAX_START="<<<"
+KERNEL_SYNTAX_END=">>>"
+DEVICE_CODE_SYNTAX="#include \"(common.*)\""
+FUNCTION_END=");"
+SPECIAL_SUFFIX="_AUTOHOSTFUNC"
+EXTRACT_KERNEL="false"
+GLOBAL_FILE="global_kernel"
+CONFIG_SELECTION_SUFFIX="_CONFIG"
+
+check_closed() {
+    local str="$1"
+    str="${str//->}"
+    str_start="${str//[^(<\[]}"
+    str_end="${str//[^>)\]]}"
+    if [[ "${#str_start}" -eq "${#str_end}" ]]; then
+        echo "true"
+    else
+        echo "false"
+    fi
+}
+
+convert_syntax() {
+    local syntax_regex="([^<>]*)(<[^<>]*>)?<<<(.*)>>>(.*)"
+    local str="$1"
+    str=$(echo "${str}" | sed -E 's/ +/ /g')
+    # echo "str ${str}"
+    local temp=""
+    local num=0
+    local var=""
+    if [[ "${str}" =~ ${syntax_regex} ]]; then
+        content="${BASH_REMATCH[3]}"
+        # echo "// content ${content}"
+        IFS=',' read -ra par <<< "$content"
+        for variable in "${par[@]}"; do
+            if [ -n "${temp}" ]; then
+                temp="${temp},"
+            fi
+            temp="${temp}${variable}"
+            # echo "// temp ${temp}"
+            is_closed=$(check_closed "$temp")
+            if [[ "${is_closed}" = "true" ]]; then
+                # echo "// temp ${temp}"
+                num=$((num+1))
+                # the last stream is zero, use the queue from exec
+                if [[ "${num}" -eq 4 ]] && [[ "${temp// }" -eq 0 ]]; then
+                    num=$((num-1))
+                else
+                    if [ -n "${var}" ]; then
+                        var="${var},"
+                    fi
+                    var="${var}${temp}"
+                fi
+                temp=""
+            fi
+        done
+        # echo "// var ${var}"
```

Comment:
If I'm correct, all you do in the previous loop is to count the number of
variables in the list, and you reconstruct it all in `$var`? In that case,
isn't it possible to use three regex on `$content` to cleanup the annoying
parts before counting? Like: